### PR TITLE
tyck: add positivity check

### PR DESCRIPTION
**Pull request by tsao-chi**
fix https://github.com/aya-prover/aya-dev/issues/838

### CODE REVIEW

**Comment by ice1000** [comment]
Cool!!!!

File: base/src/main/java/org/aya/tyck/StmtTycker.java [hunk 1]
```diff
@@ -228,7 +230,17 @@ public void tyckHeader(@NotNull Decl decl, @NotNull ExprTycker tycker) {
         assert structSig != null;
         var structLvl = structSig.result();
         var tele = tele(tycker, field.telescope, structLvl.isProp() ? null : structLvl);
+
+        if (!piPositivity(tele, structRef)) {
+          reporter.report(new PositivityError(field.sourcePos(), field, structRef.concrete));
+        }
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/StmtTycker.java:236]: For this part I think it's unnecessary. The tele of the struct cannot refer to the struct itself. We already have error report about that IIRC

**Review by ice1000 — COMMENTED** [review]
bors r+

**Review by ice1000 — APPROVED** [review]

File: base/src/test/resources/failure/tyck/positivity.aya.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+In file $FILE:4:4 ->
+
+  2 │   
+  3 │   open data Cat : Set
+  4 │     | meow : (Cat → Dead) → Cat
+    │       ╰──╯
+
+Error: Positivity check failed for | meow : (Cat → Dead) → Cat -> data Cat : Set
+       0
+         | meow : (Cat → Dead) → Cat
```
Comment (imkiva) [inline base/src/test/resources/failure/tyck/positivity.aya.txt:10]: I really don't understand this message....

File: base/src/main/java/org/aya/tyck/error/PositivityError.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  @Override @NotNull SourcePos sourcePos,
+  @NotNull AyaDocile what,
+  @NotNull AyaDocile result
+) implements TyckError {
+  @Override public @NotNull Doc describe(@NotNull PrettierOptions options) {
+    return Doc.sep(
+      Doc.english("Positivity check failed for"),
+      what.toDoc(options),
+      Doc.english("->"),
+      result.toDoc(options));
```
Comment (imkiva) [inline base/src/main/java/org/aya/tyck/error/PositivityError.java:34]: See other term printing problems for how to construct a readable doc.

File: base/src/test/resources/failure/tyck/positivity.aya.txt (hunk 2 above)
Comment (ice1000) [reply base/src/test/resources/failure/tyck/positivity.aya.txt:10]: I agree, it's very messy.

**Pull request merged**